src/storage/v2/storage.cpp

```diff
@@ -50,6 +52,92 @@ using OOMExceptionEnabler = utils::MemoryTracker::OutOfMemoryExceptionEnabler;
 
 namespace {
 inline constexpr uint16_t kEpochHistoryRetention = 1000;
+
+const std::string kReplicaName = "replica_name";
+const std::string kIpAddress = "replica_ip_address";
+const std::string kPort = "replica_port";
+const std::string kSyncMode = "replica_sync_mod";
+const std::string kTimeout = "replica_timeout";
+const std::string kCheckFrequency = "replica_check_frequency";
+const std::string kSSLKeyFile = "replica_ssl_key_file";
+const std::string kSSLCertFile = "replica_ssl_cert_file";
+
+struct ReplicaStatus {
+  std::string name;
+  std::string ip_address;
+  uint16_t port;
+  storage::replication::ReplicationMode sync_mode;
+  std::optional<double> timeout;
+  std::chrono::seconds replica_check_frequency;
+  std::optional<replication::ReplicationClientConfig::SSL> ssl;
+};
+
+nlohmann::json replica_status_to_json(ReplicaStatus &&status) {
+  auto data = nlohmann::json::object();
+
+  data[kReplicaName] = std::move(status.name);
+  data[kIpAddress] = std::move(status.ip_address);
+  data[kPort] = status.port;
+  data[kSyncMode] = status.sync_mode;
+
+  if (status.timeout.has_value()) {
+    data[kTimeout] = *status.timeout;
+  } else {
+    data[kTimeout] = nullptr;
+  }
+
+  data[kCheckFrequency] = status.replica_check_frequency.count();
+
+  if (status.ssl.has_value()) {
+    data[kSSLKeyFile] = std::move(status.ssl->key_file);
+    data[kSSLCertFile] = std::move(status.ssl->cert_file);
+  } else {
+    data[kSSLKeyFile] = nullptr;
+    data[kSSLCertFile] = nullptr;
+  }
+
+  return data;
+}
+
+std::optional<ReplicaStatus> json_to_replica_status(nlohmann::json &&data) {
+  ReplicaStatus replica_status;
+
+  const auto get_failed_message = [](const std::string_view message, const std::string_view nested_message) {
+    return fmt::format("Failed to deserialize replica's configuration: {} : {}", message, nested_message);
+  };
+
+  try {
+    data.at(kReplicaName).get_to(replica_status.name);
+    data.at(kIpAddress).get_to(replica_status.ip_address);
+    data.at(kPort).get_to(replica_status.port);
+    data.at(kSyncMode).get_to(replica_status.sync_mode);
+
+    if (const auto &timeout = data.at(kTimeout); !timeout.is_null()) {
+      replica_status.timeout = timeout.get<typename decltype(replica_status.timeout)::value_type>();
+    }
+
+    replica_status.replica_check_frequency = std::chrono::seconds(data.at(kCheckFrequency));
+
+    const auto &key_file = data.at(kSSLKeyFile);
+    const auto &cert_file = data.at(kSSLCertFile);
+
+    MG_ASSERT(key_file.is_null() == cert_file.is_null());
+
+    if (!key_file.is_null()) {
+      replica_status.ssl = replication::ReplicationClientConfig::SSL{};
+      replica_status.ssl->key_file = key_file.get<typename decltype(replica_status.ssl->key_file)::value_type>();
+      replica_status.ssl->cert_file = cert_file.get<typename decltype(replica_status.ssl->cert_file)::value_type>();
```

Comment:
Ohh, I think e2e tests are only passing because we are not using SSL for replicas. Maybe consider extracting the logic for persisting/reloading replicas into a class that can be well unit tested without spinning up multiple processes. If that is not suitable in your opinion, than at least make the json <-> ReplicaStatus conversion functions public so they can be tested separately.

Reply:
> This can be rewritten using `get_to` in my opinion, so we can avoid the `decltype` part.
> 
> Also, what does `typename decltype(replica_status.ssl->key_file)::value_type` mean? As far as I understodd:
> 
>     1. `decltype(replica_status.ssl)` is `std::optional<replication::ReplicationClientConfig::SSL>`
> 
>     2. Thus `decltype(replica_status.ssl->key_file)` is `std::string`
> 
>     3. Thus `decltype(replica_status.ssl->cert_file)::value_type` is `std::string::value_type`, which is basically some kind of a character
> 
> 
> So my question is what did I miss? Of if I didn't miss anything, then how does this work?

You're right here. Originally they were both optional but I have changed it to directly use `replication::ReplicationClientConfig::SSL` afterwards.

Reply:
> Ohh, I think e2e tests are only passing because we are not using SSL for replicas. Maybe consider extracting the logic for persisting/reloading replicas into a class that can be well unit tested without spinning up multiple processes. If that is not suitable in your opinion, than at least make the json <-> ReplicaStatus conversion functions public so they can be tested separately.

The logic to parse has been moved to a separate file and is unit-tested: `src/storage/v2/replication/replication_persistance_helper.hpp` and `tests/unit/replication_persistance_helper.cpp`.
I have also added two unit-tests in `tests/unit/storage_v2_replication.cpp` that cover the registration and "un-registration" of replicas.